File: lmfetch/rankers/embedding.py
"""Embedding-based semantic ranker using ai-query."""

import asyncio
import hashlib
import json
import os
from pathlib import Path

from ..chunkers.base import Chunk
from .base import Ranker, ScoredChunk
from ai_query import embed_many, google, openai


def cosine_similarity(a: list[float], b: list[float]) -> float:
    dot = sum(x * y for x, y in zip(a, b))
    norm_a = sum(x * x for x in a) ** 0.5
    norm_b = sum(x * x for x in b) ** 0.5
    if norm_a == 0 or norm_b == 0:
        return 0.0
    return dot / (norm_a * norm_b)


class EmbeddingCache:
    def __init__(self, cache_dir: Path | None = None):
        self.cache_dir = cache_dir or Path.home() / ".cache" / "lmfetch" / "embeddings"
        self.cache_dir.mkdir(parents=True, exist_ok=True)
        self._memory_cache: dict[str, list[float]] = {}

    def _hash_text(self, text: str) -> str:
        return hashlib.sha256(text.encode()).hexdigest()[:16]

    def get(self, text: str) -> list[float] | None:
        key = self._hash_text(text)
        if key in self._memory_cache:
            return self._memory_cache[key]

        cache_file = self.cache_dir / f"{key}.json"
        if cache_file.exists():
            try:
                embedding = json.loads(cache_file.read_text())
                self._memory_cache[key] = embedding
                return embedding
            except Exception:
                pass
        return None

    def set(self, text: str, embedding: list[float]):
        key = self._hash_text(text)
        self._memory_cache[key] = embedding
        cache_file = self.cache_dir / f"{key}.json"
        try:
            cache_file.write_text(json.dumps(embedding))
        except Exception:
            pass


class EmbeddingRanker(Ranker):
    def __init__(
        self,
        model: str = "text-embedding-005",
        batch_size: int = 20,
        provider_options: dict | None = None,
    ):
        self.model = model
        self.batch_size = batch_size
        self.provider_options = provider_options or {}
        self.cache = EmbeddingCache()

    async def _embed_texts(self, texts: list[str]) -> list[list[float]]:
        # Check cache first
        results: list[list[float] | None] = []
        uncached_indices: list[int] = []
        uncached_texts: list[str] = []

        for i, text in enumerate(texts):
            cached = self.cache.get(text)
            if cached:
                results.append(cached)
            else:
                results.append(None)
                uncached_indices.append(i)
                uncached_texts.append(text[:8000])  # Truncate for embedding

        # Embed uncached texts in batches
        if uncached_texts:
            for batch_start in range(0, len(uncached_texts), self.batch_size):
                batch = uncached_texts[batch_start : batch_start + self.batch_size]
                try:
                    # Determine provider and model object
                    model_obj = google.embedding(self.model)

                    response = await embed_many(
                        model=model_obj,
                        values=batch,
                        provider_options=self.provider_options,
                    )
                    print(response.embeddings)
                    for j, embedding in enumerate(response.embeddings):
                        idx = uncached_indices[batch_start + j]
                        results[idx] = embedding
                        self.cache.set(texts[idx], embedding)
                except Exception:
                    # If embedding fails, leave as None (will be filtered out)
                    pass

        return [r if r else [] for r in results]

    async def rank(self, query: str, chunks: list[Chunk]) -> list[ScoredChunk]:
        if not chunks:
            return []

        # Get embeddings
        texts = [query] + [c.content[:2000] for c in chunks]
        embeddings = await self._embed_texts(texts)

        if not embeddings or not embeddings[0]:
            # Fallback to keyword ranking
            from .keyword import KeywordRanker
            return await KeywordRanker().rank(query, chunks)

        query_embedding = embeddings[0]
        chunk_embeddings = embeddings[1:]

        # Score by cosine similarity
        scored = []
        for chunk, embedding in zip(chunks, chunk_embeddings):
            if embedding:
                score = cosine_similarity(query_embedding, embedding)
            else:
                score = 0.0
            scored.append(ScoredChunk(chunk=chunk, score=score))

        scored.sort(key=lambda x: x.score, reverse=True)
        return scored

File: lmfetch-bun/src/rankers/embedding.ts
/**
 * Semantic embedding ranker using AI SDK - OPTIMIZED
 */
import { embedMany } from "ai";
import { google } from "@ai-sdk/google";
import { openai } from "@ai-sdk/openai";
import { join } from "path";
import { mkdir, readFile, writeFile } from "fs/promises";
import { getCacheDir, cosineSimilarity, retry } from "../utils";
import { hashContent } from "../utils";
import type { Chunk, ScoredChunk } from "../chunkers/types";
import type { Ranker } from "./types";

const EMBEDDINGS_DIR = join(getCacheDir(), "embeddings");
const BATCH_SIZE = 100; // Increased from 20 for better performance
const MAX_EMBED_CHARS = 8000;
const EMBED_TIMEOUT = 30000; // 30 second timeout per batch

// Default to Google's text-embedding-004
const DEFAULT_PROVIDER = "google";

export class EmbeddingRanker implements Ranker {
  private cache: Map<string, number[]> = new Map();
  private provider: "google" | "openai";

  constructor(provider: "google" | "openai" = DEFAULT_PROVIDER) {
    this.provider = provider;
  }

  /**
   * Get the embedding model based on provider
   */
  private getModel() {
    if (this.provider === "openai") {
      return openai.textEmbeddingModel("text-embedding-3-small");
    }
    return google.textEmbeddingModel("text-embedding-004");
  }

  /**
   * Create context-enriched text for embedding
   */
  private enrichWithContext(chunk: Chunk): string {
    const parts: string[] = [];

    // Add file path context
    parts.push(`File: ${chunk.relativePath}`);

    // Add name if available
    if (chunk.name) {
      parts.push(`${chunk.type}: ${chunk.name}`);
    }

    // Add content (truncated)
    const content = chunk.content.slice(0, MAX_EMBED_CHARS);
    parts.push(content);

    return parts.join("\n");
  }

  /**
   * Load embedding from disk cache
   */
  private async loadFromDisk(key: string): Promise<number[] | null> {
    try {
      const path = join(EMBEDDINGS_DIR, `${key}.json`);
      const content = await readFile(path, "utf-8");
      return JSON.parse(content);
    } catch {
      return null;
    }
  }

  /**
   * Save embedding to disk cache (async, fire and forget)
   */
  private saveToDisk(key: string, embedding: number[]): void {
    // Fire and forget - don't await
    mkdir(EMBEDDINGS_DIR, { recursive: true })
      .then(() => {
        const path = join(EMBEDDINGS_DIR, `${key}.json`);
        return writeFile(path, JSON.stringify(embedding));
      })
      .catch(() => {
        // Ignore cache write errors
      });
  }

  /**
   * Embed multiple texts with aggressive caching and batching
   */
  private async embedTexts(texts: string[]): Promise<number[][]> {
    const results: (number[] | null)[] = new Array(texts.length).fill(null);
    const toEmbed: { index: number; text: string; key: string }[] = [];

    // Check cache first (memory + disk in parallel)
    await Promise.all(
      texts.map(async (text, i) => {
        const key = hashContent(text);

        // Check memory cache
        if (this.cache.has(key)) {
          results[i] = this.cache.get(key)!;
          return;
        }

        // Check disk cache
        const cached = await this.loadFromDisk(key);
        if (cached) {
          this.cache.set(key, cached);
          results[i] = cached;
          return;
        }

        toEmbed.push({ index: i, text, key });
      })
    );

    // Embed uncached texts in large batches
    if (toEmbed.length > 0) {
      // Split into batches
      for (let i = 0; i < toEmbed.length; i += BATCH_SIZE) {
        const batchItems = toEmbed.slice(i, i + BATCH_SIZE);
        const batchTexts = batchItems.map((item) => item.text);

        try {
          // Embed with timeout
          const embeddings = await Promise.race([
            this.embedBatch(batchTexts),
            new Promise<number[][]>((_, reject) =>
              setTimeout(() => reject(new Error("Embedding timeout")), EMBED_TIMEOUT)
            ),
          ]);

          // Store results
          for (let j = 0; j < batchItems.length; j++) {
            const { index, key } = batchItems[j];
            const embedding = embeddings[j];

            // Cache in memory
            this.cache.set(key, embedding);
            // Save to disk async
            this.saveToDisk(key, embedding);

            results[index] = embedding;
          }
        } catch (error) {
          console.error(`Embedding batch failed:`, error);
          // Fill failed embeddings with zeros
          for (const { index } of batchItems) {
            results[index] = new Array(768).fill(0); // Standard embedding size
          }
        }
      }
    }

    return results as number[][];
  }

  /**
   * Embed a single batch with the API
   */
  private async embedBatch(texts: string[]): Promise<number[][]> {
    const { embeddings } = await retry(
      async () => {
        return await embedMany({
          model: this.getModel(),
          values: texts,
        });
      },
      { retries: 2, delay: 500, backoff: 1.5 }
    );

    return embeddings;
  }

  /**
   * Embed a single query
   */
  async embedQuery(query: string): Promise<number[]> {
    const key = hashContent(query);

    // Check cache
    if (this.cache.has(key)) {
      return this.cache.get(key)!;
    }

    // Check disk cache
    const cached = await this.loadFromDisk(key);
    if (cached) {
      this.cache.set(key, cached);
      return cached;
    }

    // Embed the query
    const [embedding] = await this.embedBatch([query]);
    this.cache.set(key, embedding);
    this.saveToDisk(key, embedding);

    return embedding;
  }

  async rank(chunks: Chunk[], query: string): Promise<ScoredChunk[]> {
    // Embed query
    const queryEmbedding = await this.embedQuery(query);

    // Embed all chunks
    const texts = chunks.map((c) => this.enrichWithContext(c));
    const embeddings = await this.embedTexts(texts);

    // Calculate similarity scores
    const scored = chunks.map((chunk, i) => ({
      ...chunk,
      score: cosineSimilarity(queryEmbedding, embeddings[i]),
    }));

    // Sort by score descending
    scored.sort((a, b) => b.score - a.score);

    return scored;
  }
}
